packages/upgrade/src/actions/install.ts

```diff
@@ -0,0 +1,125 @@
+import { shell } from '../shell.js';
+import { random, sleep } from '@astrojs/cli-kit/utils';
+
+export async function install(
+	ctx: Pick<Context, 'version' | 'packages' | 'packageManager' | 'prompt' | 'dryRun' | 'exit' | 'cwd'>
+) {
+	await banner();
+	log('')
+	const { current, dependencies, devDependencies } = filterPackages(ctx);
+	const toInstall = [...dependencies, ...devDependencies];
+	for (const packageInfo of current.sort(sortPackages)) {
+		const tag = /^\d/.test(packageInfo.targetVersion) ? packageInfo.targetVersion : packageInfo.targetVersion.slice(1)
+		await info(`${packageInfo.name}`, `is up to date on`, `v${tag}`)
+		await sleep(random(50, 150));
+	}
+	if (toInstall.length === 0 && !ctx.dryRun) {
+		log('')
+		await success(random(celebrations), random(done));
+		return;
+	}
+	const majors: PackageInfo[] = []
+	for (const packageInfo of [...dependencies, ...devDependencies].sort(sortPackages)) {
+		const word = ctx.dryRun ? 'can' : 'will';
+		await upgrade(packageInfo, `${word} be updated to`)
+		if (packageInfo.isMajor) {
+			majors.push(packageInfo)
+		}
+	}
+	if (majors.length > 0) {
+		const { proceed } = await ctx.prompt({
+			name: 'proceed',
+			type: 'confirm',
+			label: title('wait'),
+			message: `${pluralize(['One package has', 'Some packages have'], majors.length)} breaking changes. Continue?`,
+			initial: true,
+		});
+		if (!proceed) {
+			return ctx.exit(0);
+		}
+		
+		log('');
+		
+		await warn('check', `Be sure to follow the ${pluralize('CHANGELOG', majors.length)}.`);
+		for (const pkg of majors.sort(sortPackages)) {
+			await changelog(pkg.name, pkg.changelogTitle!, pkg.changelogURL!);
+		}
+	}
+
+	log('')
+	if (ctx.dryRun) {
+		await info('--dry-run', `Skipping dependency installation`);
+	} else {
+		await runInstallCommand(ctx, dependencies, devDependencies);
+	}
+}
+
+function filterPackages(ctx: Pick<Context, 'packages'>) {
+	const current: PackageInfo[] = [];
+	const dependencies: PackageInfo[] = [];
+	const devDependencies: PackageInfo[] = [];
+	for (const packageInfo of ctx.packages) {
+		const { currentVersion, targetVersion, isDevDependency } = packageInfo;
+		// Remove prefix from `currentVersion` before comparing
+		if (currentVersion.replace(/^\D+/, '') === targetVersion) {
+			current.push(packageInfo);
+		} else {
+			const arr = isDevDependency ? devDependencies : dependencies;
+			arr.push(packageInfo);
+		}
+	}
+	return { current, dependencies, devDependencies }
+}
+
+function sortPackages(a: PackageInfo, b: PackageInfo): number {
+	if (a.isMajor && !b.isMajor) return 1;
+	if (b.isMajor && !a.isMajor) return -1;
+	if (a.name === 'astro') return -1;
+	if (b.name === 'astro') return 1;
+	if (a.name.startsWith('@astrojs') && !b.name.startsWith('@astrojs')) return -1;
+	if (b.name.startsWith('@astrojs') && !a.name.startsWith('@astrojs')) return 1;
+	return a.name.localeCompare(b.name);
+}
```

Comment:
Does this function **change** the order of the final `package.json` (in case there are some FS operations)?

Reply:
Great question! Nope, this function is only for display purposes in this CLI. Package managers will respect the existing order of dependencies in the `package.json` file. Left a comment to this effect in [27d9409](https://github.com/withastro/astro/pull/8525/commits/27d940953b7e1638c81c86917906e60537edeaa8).